From 5b1a20b44503f5f7c2303ce34fa8c36c4d32f18a Mon Sep 17 00:00:00 2001
From: Andy Ross <andrew.j.ross@intel.com>
Date: Fri, 12 Sep 2014 15:38:29 -0700
Subject: [PATCH 100/117] ov5693: Allow single-GPIO operation

The datasheet allows that one of XSHUTDN and RESETB can be tied high
and the other used as a single enable GPIO, and some modules in the
wild operate that way.  Allow the platform layer to provide only one
GPIO, and ignore errors from the second.

Issue: IRDA-748
Change-Id: I3468551a8708423db7c1b5a220445c0fd016ba34
Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
---
 .../external_drivers/camera/drivers/media/i2c/ov5693.c   | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693.c
index 2fcbb71..094fda8 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693.c
@@ -979,15 +979,15 @@ static int gpio_ctrl(struct v4l2_subdev *sd, bool flag)
 	if (dev->platform_data->gpio_ctrl)
 		return dev->platform_data->gpio_ctrl(sd, flag);
 
-	/* Note 1: CTS driver would, on flag==1 pulse reset low for
-	 * 60ms first before setting it high if CONFIG_BOARD_CTP is
-	 * set.  On which hardware is that needed? */
-
-	/* Note 2: the GPIO order is asymmetric: always RESET#
-	 * before PWDN# when turning it on or off. */
-
 	ret = dev->platform_data->gpio0_ctrl(sd, flag);
-	ret |= dev->platform_data->gpio1_ctrl(sd, flag);
+
+	/* The OV5693 has two enable inputs: XSHUTDN and RESETB, both
+	 * are active low, both must be high to enable the device.
+	 * And they can be enabled in either order.  The datasheet
+	 * even suggests that one be tied high, and some modules do
+	 * that.  Basically: allow the second GPIO to be missing in
+	 * the DSDT and ignore an error here. */
+	dev->platform_data->gpio1_ctrl(sd, flag);
 
 	return ret;
 }
-- 
1.9.3

